src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs

```diff
@@ -278,13 +278,57 @@ protected Task ResolveAuthorityAsync()
             string scopes = GetOverridenScopes(AuthenticationRequestParameters.Scope).AsSingleString();
             var tokenClient = new TokenClient(AuthenticationRequestParameters);
 
+            var CCSHeader = GetCCSHeader(additionalBodyParameters);
+            if (CCSHeader != null)
+            {
+                tokenClient.AddHeaderToClient(CCSHeader.Item1, CCSHeader.Item2);
+            }
+
             return tokenClient.SendTokenRequestAsync(
                 additionalBodyParameters,
                 scopes,
                 tokenEndpoint,
                 cancellationToken);
         }
 
+        //The CCS header is used by the CCS service to help route requests to resources in Azure during requests to speed up authentication.
+        //It consists of either the ObjectId.TenantId or the upn of the account signign in.
+        //See https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/2525
+        protected virtual Tuple<string, string> GetCCSHeader(IDictionary<string, string> additionalBodyParameters)
```

Comment:
nit: use `KeyValuePair<string, string>` instead of the tuple. The KeyValuePair is the natural structure that goes into a dictionary. 

nit: tuples are very cool, but sadly we need to use C# 8 to use them. We can't because .NET 4.5 support, so our use of Tuples is not great, t.Item1, t.Item2 etc. Hence KeyValuePair struct here might be a better fit.